### DISCUSSION TITLE
v0.4.0 - ZeroShotClassificationExplainer, Custom Labels for SequenceClassificationExplainer

### DISCUSSION THREAD

**Discussion by cdpierse** [cdpierse/transformers-interpret#48] [General]
# Zero-Shot Classification Explainer (#19,  #40)
This release introduces a new explainer `ZeroShotClassificationExplainer` that allows for word attributions to be calculated in a zero shot manner with the appropriate NLI type models.  To achieve this we implement zero shot exactly how the hugging face team did based off the following [paper](https://arxiv.org/abs/1909.00161). A list of compatible models with this explainer can be found at the transformers [model hub](https://huggingface.co/models?pipeline_tag=zero-shot-classification).

Below is an example of instantiating and calling the `ZeroShotClassificationExplainer`:

```python
from transformers import AutoModelForSequenceClassification, AutoTokenizer
from transformers_interpret import ZeroShotClassificationExplainer

tokenizer = AutoTokenizer.from_pretrained("facebook/bart-large-mnli")

model = AutoModelForSequenceClassification.from_pretrained("facebook/bart-large-mnli")


zero_shot_explainer = ZeroShotClassificationExplainer(model, tokenizer)


word_attributions = zero_shot_explainer(
    "Today apple released the new Macbook showing off a range of new features found in the proprietary silicon chip computer. ",
    labels = ["finance", "technology", "sports"], # any labels you want can be passed, even a list with one label
)
```

Which will return the following list of tuples:

```python
>>> word_attributions
[('<s>', 0.0),
 ('Today', 0.0),
 ('apple', 0.22505152647747717),
 ('released', -0.16164146624851905),
 ('the', 0.5026975657258089),
 ('new', 0.052589263167955536),
 ('Mac', 0.2528325960993759),
 ('book', -0.06445090203729663),
 ('showing', -0.21204922293777534),
 ('off', 0.06319714817612732),
 ('a', 0.032048012090796815),
 ('range', 0.08553079346908955),
 ('of', 0.1409201107994034),
 ('new', 0.0515261917112576),
 ('features', -0.09656406466213506),
 ('found', 0.02336613296843605),
 ('in', -0.0011649894272190678),
 ('the', 0.14229640664777807),
 ('proprietary', -0.23169065661847646),
 ('silicon', 0.5963924257008087),
 ('chip', -0.19908474233975806),
 ('computer', 0.030620295844734646),
 ('.', 0.1995076958535378)]
```

We can find out which label was predicted with:

```python
>>> zero_shot_explainer.predicted_label   
'technology (entailment)'
```
For the `ZeroShotClassificationExplainer` the visualize() method returns a table similar to the `SequenceClassificationExplainer`.

```python
zero_shot_explainer.visualize("zero_shot.html")
```
<img width="1155" alt="zero_shot_example" src="https://user-images.githubusercontent.com/8831892/119542425-6d7bda00-bd87-11eb-962e-eb6e087abf3d.png">

# Custom Labels For Sequence Classification -  @lalitpagaria  (#25, #41 ) 

This contribution by @lalitpagaria adds the ability to add custom class labels that replace the default labels originally found in the models's config.

This is a very useful addition as it is quite common for popular trained models to have not set the label names resulting in labels that look like `"LABEL_0", "LABEL_1",...`. This can make the sequence classification explainer visualization particularly hard to understand and not very readable. 

Custom labels are passed to the `SequenceClassificationExplainers`'s constructor and the number of labels passed must equal the number of already existing labels:

```python
seq_explainer = SequenceClassificationExplainer(
        DISTILBERT_MODEL, DISTILBERT_TOKENIZER, custom_labels=["sad", "happy"]
    )
```

Now the class at the 0th index corresponds to the label "sad" and the class at the 1st index corresponds to "happy".

This is a really nice addition that makes Transformers Interpret more usable with a whole range sequence classification models that don't have labels set. Thanks @lalitpagaria . 

# General Cleanup and Housekeeping
* Cleaned up a number of flake8 reported linting errors
* Improved on the docstring for QA explainer (more to come however) but the QA explainer still needs some finalization
* Added some increased testing coverage
* Added a contribution guideline

<hr /><em>This discussion was created from the release <a href='https://github.com/cdpierse/transformers-interpret/releases/tag/v0.4.0'>v0.4.0 - ZeroShotClassificationExplainer, Custom Labels for SequenceClassificationExplainer</a>.</em>

**Comment by lalitpagaria** [comment]
Awesome thank you @cdpierse 

One query, is it possible to get `word_attributions` against all `labels`.
For example, let's say what if user want to see explaination of top 2 or top 3 labels?
Also zero-shot supports `multi_class` which evaluates prediction independently against each labels in the list.

WDYT about these enhancements?

**Reply by cdpierse** [reply]
Hi @lalitpagaria,

That's a nice suggestion I don't think it should be _too_ difficult to make that change, which is to do it almost in a similar way to the QA explainer and have multiple data records on display. I'll look into seeing I can get that done in the next few days and make it a patch to current release. 

As for the `multi_class` feature it's mostly just a visual aid added on top in the HF pipeline, it's essentially the difference between using softmax versus sigmoid on the label outputs, it wouldn't change which class is the top predicted one. When I'm implementing the multiple  labels I'll set it so that the prediction probs are calculated with sigmoid instead of softmax thus ensuring the appearance of independence.

**Reply by lalitpagaria** [reply]
Thanks @cdpierse for considering this.

> That's a nice suggestion I don't think it should be too difficult to make that change, which is to do it almost in a similar way to the QA explainer and have multiple data records on display. I'll look into seeing I can get that done in the next few days and make it a patch to current release.

**Reply by lalitpagaria** [reply]
@cdpierse Sorry to nudge you. We are planning to add this to Obsei. https://github.com/lalitpagaria/obsei/issues/78

It can be implemented in hackish way (inside Obsei) but better to do properly, hence we will wait 🙂 -

> to get word_attributions against all labels.

Please let me know if you need help in this regard, I can raise PR.

**Reply by cdpierse** [reply]
Hi @lalitpagaria , no problem. This PR is underway, there was a bit more work to it than I anticipated but I have a working version now so I'm just working on tests and changing the docs to match the new way it work.

**Reply by cdpierse** [reply]
@lalitpagaria  I haven't written up the new release just yet but version 0.5.0 of the package is out now and has the reworked zero shot explainer included. It now calculates and visualizes attributions for all the proposed labels, hopefully this will be useful to your use case!
